File: quest-sys/src/calculations.cpp
#include "bindings.h"

namespace quest_sys {
    // Calculations
    qreal calcExpecPauliStr(Qureg const& qureg, PauliStr const& str) {
        return ::calcExpecPauliStr(qureg, str);
    }
    qreal calcExpecPauliStrSum(Qureg const& qureg, PauliStrSum const& sum) {
        return ::calcExpecPauliStrSum(qureg, sum);
    }
    qreal calcExpecFullStateDiagMatr(Qureg const& qureg, FullStateDiagMatr const& matr) {
        return ::calcExpecFullStateDiagMatr(qureg, matr);
    }
    qreal calcExpecFullStateDiagMatrPower(Qureg const& qureg, FullStateDiagMatr const& matr, Quest_Complex exponent) {
        return ::calcExpecFullStateDiagMatrPower(qureg, matr, exponent);
    }

    qreal calcTotalProb(Qureg const& qureg) {
        return ::calcTotalProb(qureg);
    }
    qreal calcProbOfBasisState(Qureg const& qureg , qindex index) {
        return ::calcProbOfBasisState(qureg, index);
    }
    qreal calcProbOfQubitOutcome(Qureg const& qureg, int qubit, int outcome) {
        return ::calcProbOfQubitOutcome(qureg, qubit, outcome);
    }
    qreal calcProbOfMultiQubitOutcome(Qureg const& qureg, rust::Slice<int> qubits, rust::Slice<int> outcomes, int numQubits) {
        return ::calcProbOfMultiQubitOutcome(qureg, qubits.data(), outcomes.data(), numQubits);
    }
    void  calcProbsOfAllMultiQubitOutcomes(rust::Slice<qreal> outcomeProbs, Qureg const& qureg, rust::Slice<int> qubits, int numQubits) {
        return ::calcProbsOfAllMultiQubitOutcomes(outcomeProbs.data(), qureg, qubits.data(), numQubits);
    }

    qreal calcPurity(Qureg const& qureg) {
        return ::calcPurity(qureg);
    }
    qreal calcFidelity(Qureg const& qureg, Qureg const& other) {
        return ::calcFidelity(qureg, other);
    }
    qreal calcDistance(Qureg const& qureg1, Qureg const& qureg2) {
        return ::calcDistance(qureg1, qureg2);
    }

    std::unique_ptr<Qureg> calcPartialTrace(Qureg const& qureg, rust::Slice<int> traceOutQubits, int numTraceQubits) {
        return std::make_unique<Qureg>(::calcPartialTrace(qureg, traceOutQubits.data(), numTraceQubits));
    }
    std::unique_ptr<Qureg> calcReducedDensityMatrix(Qureg const& qureg, rust::Slice<int> retainQubits, int numRetainQubits) {
        return std::make_unique<Qureg>(::calcReducedDensityMatrix(qureg, retainQubits.data(), numRetainQubits));
    }
    void setQuregToPartialTrace(Qureg& out, Qureg const& in, rust::Slice<int> traceOutQubits, int numTraceQubits) {
        ::setQuregToPartialTrace(out, in, traceOutQubits.data(), numTraceQubits);
    }
    void setQuregToReducedDensityMatrix(Qureg& out, Qureg const& in, rust::Slice<int> retainQubits, int numRetainQubits) {
        ::setQuregToReducedDensityMatrix(out, in, retainQubits.data(), numRetainQubits);
    }

    Quest_Complex calcInnerProduct(Qureg const& qureg1, Qureg const& qureg2) {
        return ::calcInnerProduct(qureg1, qureg2);
    }
    Quest_Complex calcExpecNonHermitianPauliStrSum(Qureg const& qureg, PauliStrSum const& sum) {
        return ::calcExpecNonHermitianPauliStrSum(qureg, sum);
    }
    Quest_Complex calcExpecNonHermitianFullStateDiagMatr(Qureg const& qureg, FullStateDiagMatr const& matr) {
        return ::calcExpecNonHermitianFullStateDiagMatr(qureg, matr);
    }
    Quest_Complex calcExpecNonHermitianFullStateDiagMatrPower(Qureg const& qureg, FullStateDiagMatr const& matrix, Quest_Complex exponent) {
        return ::calcExpecNonHermitianFullStateDiagMatrPower(qureg, matrix, exponent);
    }
}

File: quest-sys/src/calculations.rs
use crate::types::ffi::*;

#[cxx::bridge]
pub mod ffi {

    // Types
    unsafe extern "C++" {
        // Opaque QuEST types
        type Qureg = crate::types::ffi::Qureg;
        type FullStateDiagMatr = crate::types::ffi::FullStateDiagMatr;
        type PauliStr = crate::types::ffi::PauliStr;
        type PauliStrSum = crate::types::ffi::PauliStrSum;

        // Common type
        type Quest_Complex = crate::types::ffi::Quest_Complex;
    }

    // Calculations
    #[namespace = "quest_sys"]
    unsafe extern "C++" {
        include!("calculations.hpp");

        // Expectation value calculations
        fn calcExpecPauliStr(qureg: &Qureg, str: &PauliStr) -> f64;
        fn calcExpecPauliStrSum(qureg: &Qureg, sum: &PauliStrSum) -> f64;
        fn calcExpecFullStateDiagMatr(qureg: &Qureg, matr: &FullStateDiagMatr) -> f64;
        fn calcExpecFullStateDiagMatrPower(
            qureg: &Qureg,
            matr: &FullStateDiagMatr,
            exponent: Quest_Complex,
        ) -> f64;

        // Probability calculations
        fn calcTotalProb(qureg: &Qureg) -> f64;
        fn calcProbOfBasisState(qureg: &Qureg, index: i64) -> f64;
        fn calcProbOfQubitOutcome(qureg: &Qureg, qubit: i32, outcome: i32) -> f64;
        fn calcProbOfMultiQubitOutcome(qureg: &Qureg, qubits: &[i32], outcomes: &[i32]) -> f64;
        fn calcProbsOfAllMultiQubitOutcomes(
            outcomeProbs: &mut [f64],
            qureg: &Qureg,
            qubits: &[i32],
        );

        // Purity and fidelity
        fn calcPurity(qureg: &Qureg) -> f64;
        fn calcFidelity(qureg: &Qureg, other: &Qureg) -> f64;
        fn calcDistance(qureg1: &Qureg, qureg2: &Qureg) -> f64;

        // Partial trace operations
        fn calcPartialTrace(qureg: &Qureg, traceOutQubits: &[i32]) -> UniquePtr<Qureg>;
        fn calcReducedDensityMatrix(qureg: &Qureg, retainQubits: &[i32]) -> UniquePtr<Qureg>;
        fn setQuregToPartialTrace(out: Pin<&mut Qureg>, in_: &Qureg, traceOutQubits: &[i32]);
        fn setQuregToReducedDensityMatrix(out: Pin<&mut Qureg>, in_: &Qureg, retainQubits: &[i32]);

        // Non-Hermitian expectation values
        fn calcInnerProduct(qureg1: &Qureg, qureg2: &Qureg) -> Quest_Complex;
        fn calcExpecNonHermitianPauliStrSum(qureg: &Qureg, sum: &PauliStrSum) -> Quest_Complex;
        fn calcExpecNonHermitianFullStateDiagMatr(
            qureg: &Qureg,
            matr: &FullStateDiagMatr,
        ) -> Quest_Complex;
        fn calcExpecNonHermitianFullStateDiagMatrPower(
            qureg: &Qureg,
            matrix: &FullStateDiagMatr,
            exponent: Quest_Complex,
        ) -> Quest_Complex;
    }
}